### PR TITLE
Contract pdf

### PR DESCRIPTION
**Pull request by manojkrsaini09**

### CODE REVIEW

File: src/main/java/creativei/controller/MainController.java [hunk 1]
```diff
@@ -205,7 +205,8 @@ public void generateSendPdf(HttpServletResponse res) {
 			    Map<String,Object> ctx = new HashMap<String,Object>();
 			    ctx.put("name", "Manoj");
 			    ctx.put("lastName", "Saini");
-			    String html = templateService.generateHTML("VMDemo.vm", "templates/", ctx);
+			    String html = templateService.generateHTML("AmcContractOffer.vm", "templates/", ctx);
```
Comment (aayushKhandpur) [inline src/main/java/creativei/controller/MainController.java]: Move hardcoded content to common constants file.

File: src/main/java/creativei/service/EmailAsyncService.java [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+    private ProductService productService;
+
+    @Async
+    @Transactional
+    public void sendAmcContractToCustomer(AmcContract contract){
+        try {
+            AmcContractTemplateVO contractVO = this.prepareAmcContractTemplateVO(contract);
+            String customerMailId = contractVO.getCustomerVO().getEmail();
+            if (StringUtils.isBlank(customerMailId)) {
+                logger.error("Invalid email id.");
```
Comment (aayushKhandpur) [inline src/main/java/creativei/service/EmailAsyncService.java]: also log contract number and customer name here for better logging.

File: src/main/java/creativei/service/EmailAsyncService.java [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+                return;
+            }
+            Map<String, Object> ctx = new HashMap<String, Object>();
+            ctx.put("contractTemplateVO", contractVO);
+            String coontractHtml = templateService.generateHTML("AmcContractOffer.vm", "templates/", ctx);
```
Comment (aayushKhandpur) [inline src/main/java/creativei/service/EmailAsyncService.java]: Add constants to common constants file

File: src/main/java/creativei/service/EmailAsyncService.java [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+
+            MailObject email = new MailObject();
+            List<Recipient> recipients = new ArrayList<Recipient>();
+            Recipient recipient1 = new Recipient();
+            MailAddress mailAddress = new MailAddress(contractVO.getCustomerVO().getName(), customerMailId);
+            recipient1.setMailAddress(mailAddress);
+            recipient1.setRecipientType(Message.RecipientType.TO);
+            recipients.add(recipient1);
+            email.setRecipients(recipients);
+            email.setSubject("Offer for AMC Contract for Weighing Systems.");
```
Comment (aayushKhandpur) [inline src/main/java/creativei/service/EmailAsyncService.java]: Subjects should be picked from a constant file. Add an Email constants file.

File: src/main/java/creativei/service/EmailAsyncService.java [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+            ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+            pdfConversionService.convertHtmlToPdf(outputStream, coontractHtml);
+            byte[] bytes = outputStream.toByteArray();
+
+            MailObject email = new MailObject();
+            List<Recipient> recipients = new ArrayList<Recipient>();
+            Recipient recipient1 = new Recipient();
+            MailAddress mailAddress = new MailAddress(contractVO.getCustomerVO().getName(), customerMailId);
+            recipient1.setMailAddress(mailAddress);
+            recipient1.setRecipientType(Message.RecipientType.TO);
```
Comment (aayushKhandpur) [inline src/main/java/creativei/service/EmailAsyncService.java]: Also add cc and bcc email list. Read email ids from the properties file

File: src/main/java/creativei/service/EmailAsyncService.java [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+            attachment.setFile(ArrayUtils.toObject(bytes));
+            attachment.setFilename("Contract.pdf");
+            // added attachments in attacchment list
+            List<MailAttachment> mailAttachments = new ArrayList<MailAttachment>();
+            mailAttachments.add(attachment);
+
+            // set attachment list in attachment array
+            email.setAttachments(mailAttachments);
+            // send mail
+            boolean status = mailService.sendmail(email);
```
Comment (aayushKhandpur) [inline src/main/java/creativei/service/EmailAsyncService.java]: Where are we using this status? Should verify the value of status.

File: src/main/java/creativei/service/PdfService/ItextConversionService.java [hunk 7]
```diff
@@ -32,6 +32,7 @@ public void convertHtmlToPdf(OutputStream outputStream, String html) {
 			//write html content to pdf file and then output stream
 			XMLWorkerHelper.getInstance().parseXHtml(pdfWriter, document,new ByteArrayInputStream(html.getBytes(StandardCharsets.UTF_8))); 
 			document.close();
+			pdfWriter.close();
```
Comment (aayushKhandpur) [inline src/main/java/creativei/service/PdfService/ItextConversionService.java]: Use Finally block to close connections.

File: src/main/java/creativei/controller/MainController.java (hunk 1 above)
Comment (manojkrsaini09) [reply src/main/java/creativei/controller/MainController.java]: done

File: src/main/java/creativei/service/EmailAsyncService.java (hunk 5 above)
Comment (manojkrsaini09) [reply src/main/java/creativei/service/EmailAsyncService.java]: done

File: src/main/java/creativei/service/EmailAsyncService.java (hunk 6 above)
Comment (manojkrsaini09) [reply src/main/java/creativei/service/EmailAsyncService.java]: done

**Comment by manojkrsaini09** [comment]
all suggestions done

File: src/main/java/creativei/service/EmailAsyncService.java [hunk 8]
```diff
@@ -87,8 +90,14 @@ public void sendAmcContractToCustomer(AmcContract contract){
             recipient1.setMailAddress(mailAddress);
             recipient1.setRecipientType(Message.RecipientType.TO);
             recipients.add(recipient1);
+
+            List<Recipient> ccRecipients = CommonUtils.getCcMailRecipientList(serviceConfiguration.getCcMailIds());
+            List<Recipient> bccRecipients = CommonUtils.getBccMailRecipientList(serviceConfiguration.getBccMailIds());
```
Comment (aayushKhandpur) [inline src/main/java/creativei/service/EmailAsyncService.java]: Add null/empty check

File: src/main/java/creativei/controller/MainController.java [hunk 9]
```diff
@@ -137,7 +131,7 @@ public void sendEmail() throws Exception{
 		Recipient recipient1 = new Recipient();
 		
 		// Initialized mail address
-		MailAddress mailAddress = new MailAddress("User", mailID1);
+		MailAddress mailAddress = new MailAddress("User", CommonConstant.MAIL_ID_1);
```
Comment (aayushKhandpur) [inline src/main/java/creativei/controller/MainController.java]: Should not write this dummy recipient code explicitly. The recipient list should be passed to this method and utilised. Set the list with dummy ids or actual ids as applicable

**Comment by manojkrsaini09** [comment]
@aayushKhandpur  changes are done.

**Pull request merged**